[cpp] Marshalling Extern Types

Aidan63 opened the pull request:
Another long one, so make sure you're sitting comfortably.

Corresponding hxcpp PR: https://github.com/HaxeFoundation/hxcpp/pull/1189

# The Problem

Working with the current interop types comes with many pitfalls which are not immediately obvious, these include.
-	Externed value types (`cpp.Struct` / `cpp::Struct`) do not work as expected when captured in a closure. If you mutate one of these captured value types you are mutating a copy, not the object referenced by the variable name.
-	These struct types only really work with C structs, not C++ structs and classes which might have all sorts of _interesting_ constructors, destructors, assignment operators, etc. Structs are compared using memcmp and copies are made using memcpy, so if you have a C++ class which requires copy constructors, copy assignment, or others to work correctly, things will go wrong in ways which are painful to debug.
-	Leaking objects is very easy as destructors are not consistently called. If you place one of these structs in a class field its destructor will never be called.
-	Value types stored in classes will be initialised to zero, not have any default constructor invoked, so any default values or setup code will not be invoked.
-	It’s very easy to write code which does different things depending on the debug level and optimisations enabled by the compiler. You can easily find yourself modifying a copy of a struct in a temporary variable the compiler created, creating very hard to track down bugs.
-	Common C/C++ patterns, such as pointers to pointers, are hard to extern without lots of boilerplate and glue code wrangling.
-	All extern classes are considered native by the generator, so if you want to extern a custom `hx::Object` sub class some of the necessary GC code isn’t generated for the generational GC.
-	Many other things!

In short, the current interop handlers mostly work with basic c structs in local variables, but if you want to interop directly with C++ classes, you’re going to have a painful time in anything but the most basic cases.

If you just want to see a quick example of it all in action here's a gist which will compile on Windows and use DXGI to print out all displays connected to the first graphics adapter. DXGI uses pointers to pointers, pointers to structs, pointers to void pointers, and other C++ concepts which have been very difficult to map onto haxe in the past. But hopefully you'll agree that it looks like pretty "normal" haxe.

https://gist.github.com/Aidan63/07364c227335f02fbe50b9c9576f7544

# New Metadata

In my mind there are three categories of things you might want to extern, native value types, native pointer types, and "managed" (custom hx::Object sub classes) types. This merge introduces three new bits of metadata to represent these categories and solve the above issues.

## cpp.ValueType

Using the `@:cpp.ValueType` metadata on an extern class will cause it to be treated as a value type, so copies will be created when passing into functions, assigning to variables, etc, etc.

```haxe
@:semantics(reference)
@:cpp.ValueType({ type : 'foo', namespace : [ 'bar', 'baz' ] })
extern class Foo {
	var number : Int;

	function new() : Void;
}

function main() {
	final obj  = new Foo();
	final copy = obj; // creates a copy
}
```

I've chosen the metadata to take a struct which currently supports a `type` field for the underlying native type name and `namespace` which must be an array of string literals for the namespace the type is in. If `type` is omitted then the name of the extern class is used, if `namespace` is omitted then the type is assumed to be in the global namespace.

Using this metadata provides several guarantees old struct types. First it behaves how you'd expect when captured in a closure.

```haxe
function main() {
	final obj  = new Foo();
	final func = () -> {
		obj.number = 7;
	}

	func();

	trace(obj.number); // 7 will be printed.
}
```

Destructors are guaranteed to be called. When a value type is captured in a closure, stored in a class field, enums, anons, or any other GC objects it is "promoted" to the GC heap and the holder class its contained within registers a finaliser which will call the destructor.

Operators on the defined native type are always used, no memcmp or memcpy. Copy constructors, copy assignment, and standard equality operators are always used no matter the case.

The same sort of null checks are performed with references to these value types as standard haxe classes so you will get standard null exceptions instead of the program crashing with a memory error.

The nullability of these values types is unfortunately a bit odd... If you have a explicitly nullable TVar value type then its always promoted and can be null. But a null value type doesn't make much sense so I've disallowed value type variable declarations with no expression or with a null constant. Trying to assign a null value to a stack value type will result in a runtime null exception. Since value types in class fields and the likes are always promoted they are null if uninitialised. Ideally value type externs could have the same "not null" handling as core type abstracts, but that doesn't seem possible.

Interop with the existing pointer types is also provided as well implicit conversions to pointers on the underlying types for easier interop.

This value type metadata is also supported on extern enum abstracts. Historically externing enums have been a bit of a pain but it works pretty nicely now.

```haxe
@:semantics(reference)
@:include('colour.hpp')
@:cpp.ValueType({ type : 'colour' namespace : [ 'foo' ] })
private extern enum abstract Colour(Int) {
    @:native('red')
    var Red;

    @:native('green')
    var Green;

    @:native('blue')
    var Blue;
}
```

## cpp.PointerType

Using the `@:cpp.PointerType` metadata on an extern class will cause it to be treated as a pointer, this metadata supports the same fields as the above value type one.

Extern classes annotated with the pointer type metadata cannot have constructors as they are designed to be used with the common C/C++ pattern of free functions which allocate and free some sort of opaque pointer, or the pointer to pointer pattern.

E.g. the following native API could be externed and used as the following.

```C++
namespace foo {
	struct bar {};

	bar* allocate_bar();
	void free_bar(bar*);

	struct baz {};

	void allocate_baz(baz** pBaz);
	void free_baz(baz* baz);
}
```

```haxe
@:semantics(reference)
@:cpp.PointerType({ type : 'bar', namespace : [ 'foo' ] })
extern class Bar {}

@:semantics(reference)
@:cpp.PointerType({ type : 'baz', namespace : [ 'foo' ] })
extern class Baz {}

extern class Api {
	@:native('allocate_bar')
	function allocateBar():Bar;

	@:native('free_bar')
	function freeBar(b:Bar):Void;

	@:native('allocate_baz')
	function allocateBaz(b:haxe.extern.AsVar<Baz>):Void;

	@:native('free_baz')
	function freeBaz(b:Baz):Void;
}

function main() {
	final bar = Api.allocateBar();

	Api.freeBar(bar);

	final baz : Baz = null;

	Api.allocBaz(baz);
	Api.freeBaz(baz);
}
```

The pointer to pointer pattern which is pretty common is quite difficult to extern without custom C++ glue code, but the new pointer type externs understand this pattern and can be converted to a double pointer of the underlying type as well as a pointer to a void pointer which is also seen in many places.

Internally pointer types and value types are treated almost identically so most of the previous points apply here as well, the main exceptions being that promoted pointers don't have finalisers assigned and that null is always an allowed value.

## cpp.ManagedType

When you want to extern a custom `hx::Object` subclass then this is the metadata to use as it ensures the write barriers are generated for the generational gc. Like the above two metadata it supports the `type` and `namespace` fields.

```cpp
namespace foo {
	struct bar : public ::hx::Object {};
}
```

```haxe
@:cpp.ManagedType({ type : 'bar', namespace : [ 'foo' ] })
extern class Bar {
	function new() : Void;
}
```

In the above sample `Bar` will be generated as `::hx::ObjectPtr<bar>` in most cases.

There is one extra field to the managed type, `flags`, which is expected to be an array of identifiers and currently there is one flag, `StandardNaming`. If in C++ you use the hxcpp class declaration macro to create a custom subclass with the same naming scheme as haxe generated classes then this flag is designed for that.

```c++
HX_DECLARE_CLASS1(foo, Bar)

namespace foo {
	struct Bar_obj : public ::hx::Object {};
}
```

```haxe
@:cpp.ManagedType({ type : 'Bar', namespace : [ 'foo' ], flags : [ StandardNaming ] })
extern class Bar {
	function new() : Void;
}
```

In the above case `Bar` will be used instead of the manual `::hx::ObjectPtr` wrapping but `Bar_obj` will be used when constructing the class.

# Implementation Details and Rational

## Marshalling State

Value and pointer types are represented by the new `TCppNativeMarshalType` enum which can be in one of three states, `Stack`, `Promoted`, or `Reference`. This is the key to working around optimiser issues, capturing, and some interop features. All `TCppNativeMarshalType` fields are given the promoted state and `TVar`s can be given any three of the states. Any `TLocal` to a native marshal type is given the reference state. How TVars are given their state is important, variables allocated by the compiler are given the reference state, only variables typed by the user are given one of the stack (uncaptured) or promoted (captured or nullable) state. This means we dodge the issue with cpp.Struct where you could be operating on a copy due to compiler created variables.

TLocals of the reference state are generated with the new `cpp::marshal::Reference<T>` type which holds a pointer to a specific type and is responsible for any copying, promotion, checking, and just about everything. For the value type case it's T will be a pointer to the value type, and for pointer types will be a pointer to the pointer.

## Semantics

You are required to put the `@:semantics(reference)` metadata on an extern class when using the value or pointer type metadata, this does feel like a bit of a bodge... I was initially hoping that the `value` semantic would be what was needed, but tests start to fail when the analyser is enabled with value semantics. Maybe I'm just misinterpreting what these semantics are actually used for. With the reference semantics the tests do pass with the analyser, but from a quick glace that appears to be because many optimisations are disabled on types with that semantic meta...

## Compiler Error Numbers

There are several errors you may now get if you try and do things wrong (invalid meta expression) or which are not supported (function closures) instead of vague C++ errors. In these cases I've given them distinct error numbers in the CPPXXXX style, similar to MSVC and C# errors. I plan on documenting these since they're things users might naturally cause as opposed to internal bugs, so I thought it might be nice to give then concrete numbers for easier searching.

![image](https://github.com/user-attachments/assets/24d25ac5-1246-4c05-80ef-519d6b6f0be6)

## Scoped Metadata

I can never remember the names of the C++ specific metadata and end up sifting through the dropdown list every time, so I decided to prefix these ones with `cpp.` to make it easier.

## Metadata Style

I wanted to avoid re-using the `@:native` metadata for the extern classes as its already pretty common to do stuff like `@:native("::cpp::Struct<FooStruct>")` so by having a `type` and `namespace` field I wanted to make it clear it should be _just_ the type, nothing else. Also with this we can prefix `::` to the type / namespace combo to avoid any relative namespace resolution issues.

## Eager Promotion

Due to the very dynamic nature of hxcpp's function arguments and return types there are many places where value types which could be on the stack have to be promoted to satisfy the dynamic handling. With my callable type PR this should be solvable.

# Future Stuff

## Closures

Currently trying to use a function closure of a value or pointer type will result in a compilation error, but now that the variable promotion stuff is in place it should be possible to generate closures which capture the object to support this. Again I wouldn't want to do this until the callable change is in since that will greatly simplify things.

## Arrays and Vectors

Value types stored in contains such as arrays are in their promoted state, _not_ a contiguous chunk of memory which I originally wanted. Preserving C++ copying / construction semantics with haxe's resizable array looked to be a massive pain so I decided not to.
I do think having `haxe.ds.Vector`s of value types be contiguous should be possible and open up more easier interop possibilities.

## Un-dynamicification

Lots of the cpp std haxe types have a `Dynamic` context object which is then passed into C++ where its cast to a specific type. With the managed type meta we should be able to "un-dynamic" a lot of the standard library implementation.

# Closing

I'm sure there's stuff I've missed but this seems to be much more consistent in behaviour and nicer to use than the existing interop types, I've also added a bunch of tests on the hxcpp side to capture all sorts of edge cases I came across. I will also try and write some formal documentation for all this to encourage this over the existing types.

Aidan63 committed:
Initial class, interface, and enum separation

Aidan63 committed:
fold init stuff

Aidan63 committed:
Move interface header generation into its own module

Aidan63 committed:
reduce some duplication

Aidan63 committed:
managed interface implementation gen in its own module

Aidan63 committed:
cache dependencies in the ctx

Aidan63 committed:
move to gencpp with the other context data func

Aidan63 committed:
remove interface code from class gen modules

Aidan63 committed:
Use option for reference wildcard

Aidan63 committed:
path Map

Aidan63 committed:
Add back objc guard

Aidan63 committed:
calculate IDs up front

Aidan63 committed:
separate id lookup table

Aidan63 committed:
Object IDs module to hide cache

Aidan63 committed:
remap enum constructors

Aidan63 committed:
more interface dead code removal

Aidan63 committed:
tcpp interface type

Aidan63 committed:
name, flags, and debug level in tcpp_class

Aidan63 committed:
remove unused constructor var list return item

Aidan63 committed:
separate out header field generation

Aidan63 committed:
separate managed and native header gen functions

Aidan63 committed:
move header stuff into shared function

Aidan63 committed:
break class impl field gen function down

Aidan63 committed:
add converted fields to tcpp_class

Aidan63 committed:
Initial split for managed and native class impl

Aidan63 committed:
map for haxe and native implementations

Aidan63 committed:
use a list for haxe and native implementations

Aidan63 committed:
use tcpp_class field variables

Aidan63 committed:
field rename and meta / rtti filtering

Aidan63 committed:
create fields and default functions at class transform

Aidan63 committed:
use container flag

Aidan63 committed:
Mark and visit using variables and fix inverted container flag

Aidan63 committed:
static mark and visit use static variables list

Aidan63 committed:
print reflective fields

Aidan63 committed:
don't duplicate boot generation code

Aidan63 committed:
don't duplicate init function generation

Aidan63 committed:
don't duplicate dynamic function allocation generation

Aidan63 committed:
generate properties and __Field function

Aidan63 committed:
fold right member get fields

Aidan63 committed:
static get fields

Aidan63 committed:
shared member and static get fold functions

Aidan63 committed:
static set function uses ordered fields

Aidan63 committed:
member set uses organised fields

Aidan63 committed:
turn abstract functions into normal functions

Aidan63 committed:
cppia gen uses organised functions

Aidan63 committed:
simplify ScriptNamedFunction generation

Aidan63 committed:
bit of simplification of more cppia stuff

Aidan63 committed:
GetFields uses organised fields

Aidan63 committed:
move class retyping into retyper

Aidan63 committed:
remap interface functions

Aidan63 committed:
store interface hash

Aidan63 committed:
store meta and rtti fields in remapped interface

Aidan63 committed:
dead code removal

Aidan63 committed:
tcpp interfaces only store their functions

Aidan63 committed:
remapped enum fields rename for consistency

Aidan63 committed:
calculate interface slots ahead of time

Aidan63 committed:
remap interface arg names

Aidan63 committed:
retype the interfaces stored on retyped classes

Aidan63 committed:
retype tcpp_class supers

Aidan63 committed:
find_class_implementation uses tcpp_class

Aidan63 committed:
folds for class interface code

Aidan63 committed:
other minor cleanup

Aidan63 committed:
use flags in class generation to selectively generate some funcs

Aidan63 committed:
retype class functions

Aidan63 committed:
retype class variables

Aidan63 committed:
dedicated var and dyn function boot functions

Aidan63 committed:
dead code cleanup

Aidan63 committed:
use bindings instead of to_list

Aidan63 committed:
own implementation of of_list

Aidan63 committed:
replace another to_list

Aidan63 committed:
Fix pointer and struct wrapping

Aidan63 committed:
fix debug level not being reset between classes

Aidan63 committed:
add space after static

Aidan63 committed:
remove unused variable

Aidan63 committed:
ensure SourceFile paths are made absolute

Aidan63 committed:
some debugging

Aidan63 committed:
remove debugging prints and store class params

Aidan63 committed:
don't use already mangled name when generating the getter name in reflection

Aidan63 committed:
same for setters

Aidan63 committed:
dont use new gc references function

Aidan63 committed:
better handling of static and virtual attributes

Aidan63 committed:
generate scriptable class using tcpp class function lists

Aidan63 committed:
go back to type cant be null

Aidan63 committed:
Container flag refers to entire inheritance tree

Aidan63 committed:
allow code to easily distinguish between the different container types

Aidan63 committed:
remove accidentally tracked file

Aidan63 committed:
Merge branch 'development' into gencpp_rework_mkii

Aidan63 committed:
Merge branch 'development' into gencpp_rework_mkii

Aidan63 committed:
Merge branch 'development' into gencpp_rework_mkii

Aidan63 committed:
pass a context object around the retyper

Aidan63 committed:
closures tracked in retyper ctx

Aidan63 committed:
injection stored in retyper ctx

Aidan63 committed:
declarations track by retyper ctx

Aidan63 committed:
this handling moved into ctx

Aidan63 committed:
gc stack handled by retyper ctx

Aidan63 committed:
function return type handled by retyper ctx

Aidan63 committed:
move loop goto management into retyper ctx

Aidan63 committed:
consistent retyper ctx variable name

Aidan63 committed:
move remaining counter into the fold

Aidan63 committed:
Merge branch 'development' into gencpp_rework_mkii

Aidan63 committed:
shared string_map_of_list function

Aidan63 committed:
initial value type support

Aidan63 committed:
generic templates and correct return types

Aidan63 committed:
Boxing of captured locals and use marshalling struct helper

Aidan63 committed:
cast dynamic to a reference object

Aidan63 committed:
always use a reference object instead of sometimes directly a boxed object

Aidan63 committed:
cast variant to value type

Aidan63 committed:
error on to value type field closures

Aidan63 committed:
Use dedicated new marshal types

Aidan63 committed:
Update forgotten about old cpp reference

Aidan63 committed:
use _hx_vt as a prefix

Aidan63 committed:
support namespace field

Aidan63 committed:
omit brackets for stack constructors which have no arguments

Aidan63 committed:
Fix alignment of reference declarations in closures

Aidan63 committed:
default to creating boxed value types and deal with the few special cases

Aidan63 committed:
some nullable support

Aidan63 committed:
move auto cast into a separate file

Aidan63 committed:
retype tvar to tcpp specific type

Aidan63 committed:
retype tvar to simplify code and fix closure argument assignment

Aidan63 committed:
promote function and lambda args if captured

Aidan63 committed:
unify closure and function arg generation

Aidan63 committed:
initial attempt at assigning value types a state

Aidan63 committed:
some cleanup surrounding value type state assignment

Aidan63 committed:
remove the two variable creation system

Aidan63 committed:
implement class fields and statics

Aidan63 committed:
wrap value type arguments for new non based struct value type

Aidan63 committed:
cleanup of code and default values

Aidan63 committed:
support enums

Aidan63 committed:
treat abstract this variable as a reference and cast with a promoted type

Aidan63 committed:
pointer type interop support

Aidan63 committed:
ensure arrays of value types are boxed

Aidan63 committed:
treat nullable value types as promoted

Aidan63 committed:
default to heap construction for references

Aidan63 committed:
treat extern value type fields of extern value types as refences

Aidan63 committed:
avoid double :: when no namespace is specified

Aidan63 committed:
Merge branch 'development' into gencpp_rework_mkii

Aidan63 committed:
another pass to catch some missing :: prefixes

Aidan63 committed:
Merge branch 'gencpp_rework_mkii' into cpp-value-types

Aidan63 committed:
Initial enum support and comparison operator support

Aidan63 committed:
correct enum casting

Aidan63 committed:
handle extern value type field access and assignment correctly

Aidan63 committed:
fallback to stack allocation instead of heap

Aidan63 committed:
Allow creating reference state values types on the stack

Aidan63 committed:
require reference semantics metadata

Aidan63 committed:
Don't use the package if no namespace is provided

Aidan63 committed:
add inheritance and casting tests

Aidan63 committed:
support star of and dereference native functions

Aidan63 committed:
Initial support for pointer types

Aidan63 committed:
Add forgotten meta json change

Aidan63 committed:
make sure all boxed pointer vars are allocated

Aidan63 committed:
handle adding pointer stars once

Aidan63 committed:
Use custom marshal pointer type for pointer references

Aidan63 committed:
pointer and value specific reference cpp types

Aidan63 committed:
don't retype args with dynamic

Aidan63 committed:
bits of cleanup and specifying some types more

Aidan63 committed:
Remove need for assignment filter, now fixed at source

Aidan63 committed:
remove useless argument

Aidan63 committed:
retype null promoted pointers into boxed objects

Aidan63 committed:
inheritance casting with pointer types as well

Aidan63 committed:
Look for include meta on extern abstracts

Aidan63 committed:
don't add * to pointer type and references

Aidan63 committed:
Don't use subscript operator for pointers of marshal types

Aidan63 committed:
less code duplication and unwrap marshal type params

Aidan63 committed:
do something for cppia

Aidan63 committed:
require reference semantics on pointer types

Aidan63 committed:
Don't allow constructors on pointer types

Aidan63 committed:
Initial support for managed externs

Aidan63 committed:
first pass of the _obj handling

Aidan63 committed:
Handle static functions and variable

Aidan63 committed:
Fix incorrect template handling with standard naming

Aidan63 committed:
another standard naming fix

Aidan63 committed:
allow static fields on extern pointers

Aidan63 committed:
better static handling and helper function name consistency

Aidan63 committed:
Exclude marshal types from having debug info generated

Aidan63 committed:
rename marshal type enum for consistency

Aidan63 committed:
add managed marshal type to cppia enum

Aidan63 committed:
don't follow away marshalling enum abstracts to ensure we see meta

Aidan63 committed:
More generic extern abstract following

Aidan63 committed:
Add a bunch of errors

Aidan63 committed:
Fix error position

Aidan63 committed:
dev merge

kLabz commented:
> ## Compiler Error Numbers
>
> There are several errors you may now get if you try and do things wrong (invalid meta expression) or which are not supported (function closures) instead of vague C++ errors. In these cases I've given them distinct error numbers in the CPPXXXX style, similar to MSVC and C# errors. I plan on documenting these since they're things users might naturally cause as opposed to internal bugs, so I thought it might be nice to give then concrete numbers for easier searching.

Mostly nitpicking, but shouldn't these have a `HXCPP` rather than `CPP` prefix to avoid confusion with native errors?

kLabz commented on src-json/meta.json:
```diff
@@ -1130,5 +1130,26 @@
+		"metadata": ":cpp.ManagedType",
+		"doc": "",
+		"platforms": ["cpp"],
+		"targets": ["TClass"]
```
Would be better to move those up with other `Cpp*` metadata, and add `doc` :)

Aidan63 replied:
The other cpp meta seems to be a bit all over the place, but I've moved the new ones up to be next to a few others and filled in the doc comment.

Aidan63 committed:
Move meta up and fill in doc comment

Aidan63 commented:
> > ## Compiler Error Numbers
> > There are several errors you may now get if you try and do things wrong (invalid meta expression) or which are not supported (function closures) instead of vague C++ errors. In these cases I've given them distinct error numbers in the CPPXXXX style, similar to MSVC and C# errors. I plan on documenting these since they're things users might naturally cause as opposed to internal bugs, so I thought it might be nice to give then concrete numbers for easier searching.
> 
> Mostly nitpicking, but shouldn't these have a `HXCPP` rather than `CPP` prefix to avoid confusion with native errors?

I kept going back and forth on this. I ended up going with CPP due to the `--cpp` compiler flag and `cpp` conditional, easy enough to change though.

Aidan63 commented:
Oh, and here are some relevant issues. These are definitely fixed with the above changes.

https://github.com/HaxeFoundation/haxe/issues/11945

https://github.com/HaxeFoundation/hxcpp/issues/1099

The following issues are not directly fixed by this PR, but these new extern types do not have the issues and could be seen to supersede the old extern techniques.

https://github.com/HaxeFoundation/hxcpp/issues/1131

https://github.com/HaxeFoundation/hxcpp/issues/482

https://github.com/HaxeFoundation/hxcpp/issues/1096

kevinresol commented:
This looks nice!

I have a question about cpp.PointerType. Since it is now opaque, so how do we perform pointer operations on it such as array access or reinterpret_cast it to another cpp.PointerType?

kevinresol commented:
btw this seems to break existing code (comparing to haxe 4.3.6): 

```haxe
package gdnative.variant;
extern enum abstract Type(Type_extern) {
}
@:include("godot_cpp/variant/variant.hpp") @:native("cpp::Struct<godot::Variant::Type, cpp::EnumHandler>") extern class Type_extern {

}
```

When the type is referenced somewhere, the c++ code generated by this PR is missing the header include:

```cpp
#ifndef INCLUDED_cc2f1716c78dbbe1
#define INCLUDED_cc2f1716c78dbbe1
#include "godot_cpp/variant/variant.hpp"
#endif
```

Aidan63 commented:
If the externed pointer or value type has a subscript operator `[]` then you can have the haxe extern type implement the `haxe.extern.ArrayAccess` interface and it will go through that native operator (I think I forgot to add tests for this). Pointers which are actually arrays won't really work with the `cpp.PointerType` metadata.

Reinterpreting pointer or value types can be done with the usual haxe `cast`.

I'll take a look at that include issue, I added new meta because I didn't want to break any existing externs!

Aidan63 committed:
try a different way of dealing with extern abstracts metadata

Aidan63 committed:
actually fix include and abstract issue

Aidan63 commented:
That missing include issue should now be fixed. Was an oversight in the changes I made so that metadata on extern abstracts aren't ignored.

kevinresol commented:
I also noticed that an extra cast is inserted in the generated code for the `value` field of  ``cpp.Pointer`

was:

```cpp
( (::cpp::Pointer<  godot::ClassDBSingleton >)(this->_hx___gd->ptr) )->get_value().get_class_list()
```

now: 
```cpp
( ( godot::ClassDBSingleton)(( (::cpp::Pointer<  godot::ClassDBSingleton >)(this->_hx___gd->ptr) )->get_value()) ).get_class_list()
```

In this case it fails to compile because `~ClassDBSingleton` is protected and I got this error

```
error: variable of type 'godot::ClassDBSingleton' has protected destructor
```


Also 4.3.6 will tmp var (when needed) it as a reference `godot::ClassDBSingleton & this1 = ( (::cpp::Pointer<  godot::ClassDBSingleton >)(this->_hx___gd->ptr) )->get_value();` but now the reference is removed and became `godot::ClassDBSingleton this1`

Aidan63 commented:
I can see the extra cast as well. But it seems to be a regression in nightlies in general, not specifically this branch. I'll take a look.

kevinresol commented:
Thanks for looking into it! When it is merged, would love to see a rebase so I can test out the binary built from this PR

Aidan63 committed:
Merge branch 'development' into cpp-value-types

Aidan63 commented:
I've merged dev into this branch so new builds should have the fix.

kevinresol commented:
Okay I can compile the existing code with the haxe binary now. So I can actually start testing this out.

I just found that on the cpp side the marshalling wrapper does not seem to be compatible with cppia:

Host.hx
```haxe
@:headerCode('enum MyEnum { X = 3, Y = 27 };')
class Host {
	public static function main() {}

	public function new(type:MyEnum) {}

	static function getValue():MyEnum {
		return MyEnum.X;
	}
}

@:semantics(reference)
@:cpp.ValueType({type: 'MyEnum', namespace: []})
extern enum abstract MyEnum(Int) to Int {
	final X;
	final Y;
}
```

build.hxml
```
--main Host
--cpp bin
-cp src
-D scriptable
```


generated cpp
```cpp
::cpp::marshal::ValueType< ::MyEnum > Host_obj::getValue();

static void CPPIA_CALL __s_getValue(::hx::CppiaCtx *ctx) {
  ctx->returnInt(Host_obj::getValue());
}
```

Outcome:
cpp compiler complains that it can't convert marshalled enum to an int, for a cppia function returning the enum

```
Error: ./src/Host.cpp:84:16: error: no viable conversion from '::cpp::marshal::ValueType< ::MyEnum>' to 'int'
   84 | ctx->returnInt(Host_obj::getValue());
```

also it can't convert an int to the marshalled enum for a cppia function arg:

Aidan63 committed:
Ensure all marshal types are object returns in cppia wrapper classes

Aidan63 committed:
I knew arguments were hiding in there somewhere...

Aidan63 committed:
use proper marshalling state

Aidan63 committed:
deal with args in the return call as well

Aidan63 committed:
do same for interfaces

Aidan63 committed:
forgotten bracket

Aidan63 commented:
This should now be working, or, atleast, the marshalling tests now compile with the scriptable define which is a good start.

kevinresol commented:
hmm, I am still unable to compile the code snippet in my previous post.

Got this:

```
Error: In file included from ./src/Host.cpp:2:
In file included from /Users/kevin/haxe/haxelib/hxcpp/git/include/hxcpp.h:357:
/Users/kevin/haxe/haxelib/hxcpp/git/include/cpp/Marshal.h:618:63: error: cannot initialize a member subobject of type 'MyEnum' with an rvalue of type 'int'
  618 | inline cpp::marshal::ValueType<T>::ValueType(TArgs... args) : value( std::forward<TArgs>(args)... ) {}
      |                                                               ^      ~~~~~~~~~~~~~~~~~~~~~~~~~
./src/Host.cpp:104:65: note: in instantiation of function template specialization 'cpp::marshal::ValueType<MyEnum>::ValueType<int>' requested here
  104 |  _HX_SUPER ? ((Host_obj*)ctx->getThis())->Host_obj::__construct(ctx->getInt(sizeof(void*))) : ((Host_obj*)ctx->getThis())->__construct(ctx->getInt(sizeof(void*)));
      |                                                                 ^
./src/Host.cpp:106:51: note: in instantiation of function template specialization '__script_construct_func<false>' requested here
  106 | ::hx::ScriptFunction Host_obj::__script_construct(__script_construct_func,"vi");
      |                                                   ^
1 error generated.
 ERROR  (unknown position)

   | Error: Build failed
```

Does it happen on your side as well?

Aidan63 committed:
fix call dump

Aidan63 commented:
Yep, it was happening on my side as well. I got the marshalling tests compiling with the scriptable flag and forgot to check your original sample. There's a fair bit of duplication in the cppia support code generation which could probably do with being cleaned up...

kevinresol commented:
I guess there is another missing bit for member functions returning a value type:

```haxe
@:headerCode('enum MyEnum { X = 3, Y = 27 };')
class Host {
	public static function main() {}

	public function new(type:MyEnum) {}

	static function staticGetValue(input:MyEnum):MyEnum {
		return MyEnum.X;
	}

	function getValue(input:MyEnum):MyEnum {
		return MyEnum.X;
	}
}

@:semantics(reference)
@:cpp.ValueType({type: 'MyEnum', namespace: []})
extern enum abstract MyEnum(Int) to Int {
	final X;
	final Y;
}
```

```
Error: In file included from ./src/Host.cpp:2:
In file included from /Users/kevin/haxe/haxelib/hxcpp/git/include/hxcpp.h:357:
/Users/kevin/haxe/haxelib/hxcpp/git/include/cpp/Marshal.h:618:63: error: cannot initialize a member subobject of type 'MyEnum' with an rvalue of type 'int'
  618 | inline cpp::marshal::ValueType<T>::ValueType(TArgs... args) : value( std::forward<TArgs>(args)... ) {}
      |                                                               ^      ~~~~~~~~~~~~~~~~~~~~~~~~~
./src/Host.cpp:110:10: note: in instantiation of function template specialization 'cpp::marshal::ValueType<MyEnum>::ValueType<int>' requested here
  110 |                 return __ctx->runInt(__scriptVTable[1]);
      |                        ^
1 error generated.
 ERROR  (unknown position)
```

I guess the generated `runInt` should be `runObject`?

Aidan63 committed:
rework cppia glue generation

Aidan63 committed:
do similar cleanup for interfaces

Aidan63 commented:
yes, that one should be fixed now as well. I've done some cleanup of the cppia scaffolding generation so that should hopefully be the last of the issues (last famous words...).

Aidan63 committed:
store the interface return type as a tcpp

Aidan63 committed:
store interface args as retyped versions

kevinresol commented:
**Edited:**
The original comment was a bit messy, let me try to conclude it better.

In short I would like to know how to consistently obtain a pointer to a ValueType, and the actual value of a ValueType. For the purpose of interop-ing with native code that expects pointer of the value type and actual value of the value type. 
Because the actual promotion state in the generated native code can be different in different scenarios.


**Original:**
I noticed that following in the ValueType section:

> Interop with the existing pointer types is also provided as well implicit conversions to pointers on the underlying types for easier interop.

How does it work if I want to get the pointer to a ValueType value regardless of its promotion state?


Besides pointer, I am also curious in how to make haxe emit code that correctly dereferences a ValueReference if it decided to produce a temp var. For example, I can see that currently the following is generated for the haxe code `untyped __cpp__ ('x = {0}', p_rhs.__gd)` where `p_rhs.__gd` is a ValueType for `godot::Variant` stored in a class field `__gd`

```cpp
::cpp::marshal::ValueReference< ::godot::Variant > p_rhs1 = ( (::cpp::marshal::ValueReference< ::godot::Variant >)(p_rhs) );
return x == p_rhs1;
```

But it is not correct because I expect the actual value be used in the comparison, not the reference created by haxe.
Maybe we need some extra Haxe API for these kind of stuff?

Aidan63 commented:
There are two ways you can deal with pointers to value types. If you have a function which accepts a pointer to a type you've externed as a value type the easiest way is to just extern that function using the existing haxe extern and C++ operator conversions will take care of it. I used this method in the directx gist in the original post. E.g.

I have a value type extern for `DXGI_ADAPTER_DESC` like this.

```haxe
@:semantics(reference)
@:cpp.ValueType({ type : 'DXGI_ADAPTER_DESC' })
extern class DxgiAdapterDescription {
    @:native('VendorId')
    var vendorId : Int;

    @:native('DeviceId')
    var deviceId : Int;
}
```

and I want to use a function `IDXGIAdapter::GetDesc` which has the following signature, `HRESULT GetDesc(DXGI_ADAPTER_DESC* pDesc)`, so a pointer to my value type extern. If I simply extern that function as the following,

```haxe
@:native('GetDesc')
function getDesc(desc : DxgiAdapterDescription) : Int;
```

Then it will "Just Work" as C++ operators take care of getting a pointer to the value type passed in, no matter what the state of it is.

The alternative if you want to be a bit more explicit is to use something like `cpp.Star` or similar in that externed function signature.

```haxe
@:native('GetDesc')
function getDesc(desc : cpp.Star<DxgiAdapterDescription>) : Int;
```

When you use those existing pointer interop types with value types it will use the underlying type, not the hxcpp wrapper types. i.e. `cpp.Pointer<DxgiAdapterDescription>` becomes `::cpp::Pointer<DXGI_ADAPTER_DESC>`, not something like `::cpp::Pointer<cpp::marshal::ValueType<DXGI_ADAPTER_DESC>>`.

There are more value type interop examples in the hxcpp tests. https://github.com/HaxeFoundation/hxcpp/blob/7fcb9203ae2da9e01e82877232d7cac7122ec7bb/test/native/tests/marshalling/classes/TestValueTypeInterop.hx

For variables, only those typed by the user will be given the stack or promoted state, any internal compiler generated variables will be given the reference state. This is to avoid the sort of issues you see with`cpp.Struct` where you can end up mutating a compiler generated variable, and therefore a copy.
That reference equality can be quite deceptive, `::cpp::marshal::ValueReference` has overloaded equality operators where it will dereference the value type, it does not compare based on pointers. This means the equality operators of the native type are used.

kevinresol commented:
Thanks for the explanation, I seem to get the pointer/value part working (although I reckon it can be optimized a little bit)

For my use case I need to pass a pointer to native API, the pointer cast in Haxe works well. On the other hand I also need to be able to get the actual value for native operator overloads, which can be done via casting it to cpp.Pointer again then use its `value` field.  There are also native API that accepts a reference, but it seems fine just passing a dereferenced value to it. Though I am not very sure about the performance implication for that.

The other topic I would like to explore is about using the Haxe generated code from native side.
I noticed that ValueType used as haxe function args are generated as `cpp::marshal::Boxed<T>` and return values are generated as `cpp::marshal::ValueType<T>`. The latter looks pretty straight forward I can just get the actual value via the `value` field. The former is more compilcated, consider the following Haxe example:

```haxe
class Foo {
  var x:MyValueType;
  public function new(x:MyValueType) this.x = x;
}
```

The generated c++ will be like:

```cpp
// this->x is typed as Boxed<MyValueType>
void Foo_obj::__construct(::cpp::marshal::ValueType< MyValueType > v) {
	this->x = (::cpp::marshal::ValueReference< MyValueType >)(v);
}
```

If I understand correctly ValueType is stack allocated and I don't get how it ends up being allocated in the Heap (for ValueReference and Boxed) during the the various casts.


One more side note: extern enum abstract annotated with ValueType is not compatible with string concatenation any more. I need to do a manual int cast so this works: `trace('Hello: ${(x : Int)}`

Aidan63 commented:
> For my use case I need to pass a pointer to native API...

I'm not sure I follow this paragraph, do you have some concrete examples?

> I noticed that ValueType used as haxe function args..

Sometimes value type arguments and return types need to be boxed to work with certain haxe function "features" related to dynamic. The good news is I've worked on improving function calling in #11151 so that you only pay the dynamic cost if you're explicitly using it, so in the future value type arguments won't need to be boxed as often.

The types `::cpp::marshal::ValueType<T>` and `::cpp::marshal::Boxed<T>` can be considered the "storage" for value types, in the case of `ValueType` it will be on the stack and cleaned up when the stack unwinds, whereas `Boxed` lives in the GC heap. The third type `::cpp::marshal::ValueReference` is a sort of "universal accessor" to a value type. I didn't want to mostly duplicate all logic for value type handling in both `ValueType` and `Boxed`, so anything you do with a value type will go through that light `ValueReference` wrapper which maintains value semantics, haxe null compatibility, conversions, etc, etc.

One of the conversions is an implicit conversion to a `Boxed<T>` which is what happens in that line.

```cpp
this->x = (::cpp::marshal::ValueReference< MyValueType >)(v);
```

The `ValueReference<T>` is being assigned to the `Boxed<T>` field, so the implicit conversion function is invoked. This allocates a new `Boxed<T>` and calls the copy constructor on the dereferenced value pointer to preserve the underlying type semantics.

https://github.com/HaxeFoundation/hxcpp/blob/7fcb9203ae2da9e01e82877232d7cac7122ec7bb/include/cpp/Marshal.h#L282

Most of this should be implementation details, even if you're writing C++ glue code you shouldn't need to manually deal with the value, boxed, or reference state types (although I'm sure there are some extreme edge cases), you can just deal with standard C/C++ pointers, references, and values. Appropriate conversions should automatically happen when those type pass between haxe and C/C++.

> extern enum abstract annotated with ValueType is not compatible with string concatenation any more

I'll take a look, wonder if this is a general limit of extern enum abstracts as opposed to something wrong with the C++ stuff.

kevinresol commented:
> I'm not sure I follow this paragraph, do you have some concrete examples?

So basically I am dealing with cases where the expected type is not explicit , such as c++ operator overloads.

Let's say a class X has a op overload for value type Y:

```cpp
// c++
class X {
  bool operator==(const Y& y);
}
```

When trying to generate the comparison code in Haxe:

```haxe
// haxe
// x is instance of `extern class X {}`
// y is instance of `@:cpp.ValueType extern class Y {}`
untyped __cpp__('{0} == {1}', x, y);
```

It is not going to work because the c++ compiler would not know that it should convert `y`, which could be in any promotion state, to `const Y&` or `Y&` for choosing the operator overload.

I could manually insert a double cast via `__cpp__` but it doesn't look nice, and the native type name of `y` may not really be known here.

```haxe
// haxe
untyped __cpp__('{0} == static_cast<Y&>(static_cast<ValueReference<Y>>({1}))', x, y);
```

Hence what I was trying to ask is that we could possibly make it nicer with some sort of conversion API on the Haxe side. e.g.:

```haxe
// haxe
untyped __cpp__('{0} == {1}', x, cpp.ValueType.getReference(y));
```

and the case is similar when the actual value type value is needed. Maybe `cpp.ValueType.getValue(y)` would generate `static_cast<Y>(static_cast<ValueReference<Y>>(y))`

Not very sure if all these make sense at all, please let me know!

kevinresol commented:
I also have question regarding null. Consider the following example regarding optional function argument:

```haxe
// Foo is a @:cpp.PointerType extern class
function f(?native:Foo) {
  trace("native is null", native == null);
}

f(); // false, but I think it should be true from a Haxe point of view?
```

The generated c++ is roughly:

```cpp
void f(Boxed<Foo*> native){
	::hx::IsEq( native,Boxed<Foo*>( new Boxed_obj<Foo*> (null())) );
}

f(Boxed<Foo*>( new Boxed_obj<Foo*> (null())));
```

Aidan63 commented:
I see what you mean now, in that situation I'd write a small wrapper C++ functions, something like

```c++
bool _hx_x_eq_y(X* lhs, Y* rhs) {
    return (*lhs) == (*rhs);
}
```

and extern it to haxe with something like 

```haxe
function x_eq_y(lhs : cpp.Star<X>, rhs : cpp.Star<Y>) : Bool;
```

I'm a bit hessitant to add escape hatch functions because it opens up untyped cpp and I want to reduce the amount of that I see as much as possible. But with the above solution it does require something like an abstract to make it nice to use from haxe and requires writing a fair bit of boiler plate code.

Speaking of abstracts, before starting all this one alternative I did consider was using core type abstracts instead of extern classes as that would make all those operator overload scenarios possible.

```haxe
function main() {
    final x = new X();
    final y = new Y();

    trace(x == y);
}

@:coreType
@:notNull
@:cpp.ValueType({ type : 'X' })
abstract X {
    public function new() : Void;

    @:op(A == B)
    public function eqY(rhs:Y) : Bool;
}

@:coreType
@:notNull
@:cpp.ValueType({ type : 'Y' })
abstract Y {
    public function new() : Void;
}
```

The one main kicker with this is that there's no nice way to represent inheritance, so I opted against it in the end and went with traditional extern classes. I like this idea but I'm not sure what to do with it since a lack of inheritance support is a pretty big issue.

> I also have question regarding null

I think that was a bit of an oversight. I guess null checks involving boxed pointers should check the null-ness of the GC container and the null-ness of the held pointer. I don't think there are any scenarios where you'd want to differentiate between the two.

kevinresol commented:
It is unfortunate that the example was a little bit misleading, the main culprit is not the operator overload per se, but how the c++ compiler sees a type and select the correct construct (function call, operator, etc). So it will still happen for templated function such as:

```cpp
template<typename T>
void foo(T& ref);

template<typename T>
void bar(T* ptr);
```

In that case there is basically no way to call `foo`  with a ValueType or `bar` with a PointerType because the c++ will not recognize these marshal types as a plain reference or pointer.

For me the way to call bar with a PointerType value `v` for example:

```haxe
// works
extern static inline function bar<T>(v:T):Void 
  untyped __cpp__('bar({0})', ( v: cpp.Star<T> ));

// does not work
@:native("bar") extern static function bar<T>(v:cpp.Star<T>):Void;
```

And for c++ references unless we have something like cpp.Ampersand that does the same job as cpp.Star, we are left with doing the cast in untyped __cpp__.

Aidan63 commented:
This is possible as long as you introduce a partial template specialisation or something similar. This specialisation would just be an intermediate which is specialised to a `ValueReference` and calls the original templated function with a reference to the value type. Full sample below.

```haxe
@:cpp.ValueType
@:semantics(reference)
extern class Foo {
    var i : Int;

    function new() : Void;

    static function bar<T>(v : T) : Void;
}

@:headerNamespaceCode('
struct Foo {
    int i;

    template<class T>
    static void bar(T& v) {
        printf("base template %i\\n", v.i);

        v.i = 10;
    }

    template<class T>
    static void bar(::cpp::marshal::ValueReference<T> v) {
        printf("partial specialisation\\n");

        bar(*v.ptr);
    }
};
')
class Main {
    static function main() {
        final foo = new Foo();
        foo.i = 7;

        Foo.bar(foo);

        trace(foo.i);
    }
}
```

We trace the value of `foo.i` back in haxe to prove that we've been mutating a reference to the initially passed in value type, not a copy. In the end you should see something like this.

![image](https://github.com/user-attachments/assets/31d37d8c-8703-45dd-a4d3-e887f4a45478)

kevinresol commented:
I think the compiler should be aware of any ValueType that is created inline and immediately converted to a reference. Because in that case the reference will become invalid as the value itself is out of scope straight away.

Consider the following pseudo generated code:

```cpp
foo( (ValueReference<T>) (ValueType<T>(args)) )
```

When foo gets the reference and use it in the body, the reference is already invalid as the actual value itself is already out of scope.

In this case I believe the compiler needs to make sure the actual value can survive the `foo` call so it should generate something like:

```cpp
ValueType<T> tmp(args);
foo((ValueReference<T>)(tmp)
```

Aidan63 committed:
Merge branch 'development' into cpp-value-types

Aidan63 committed:
Merge branch 'development' into cpp-value-types

Aidan63 committed:
bloody submodules

Aidan63 committed:
require value semantics instead

Aidan63 committed:
Deal with generated temporaries which need to be value types

Also a quick fix for templated return types

Aidan63 committed:
less string maps

Aidan63 committed:
Support array access on marshaling types

Aidan63 committed:
initial templated function support

Aidan63 committed:
Guard against unresolved extern parameters

Aidan63 committed:
reinterpreting to value types

Aidan63 committed:
Fix pointer to pointer types, and add view extern

Aidan63 committed:
static function template types

Aidan63 committed:
Document some marshal functions

Aidan63 committed:
Document marshal toString functions

Aidan63 committed:
ViewExtensions convenience functions

Aidan63 committed:
view compare and marshal read / write helpers

Aidan63 commented:
I've taken a break from the coroutine mines to update this branch.

- Array access now works again by implementing the `ArrayAccess` interface.
- Templated functions can now be externed with generic functions.
- The required semantics type has been changed to `value` and fixed an issue where some generated variables can't be references.

I then got a bit carrier away and added some new types to help with externs.

### cpp.marshal.View<T>

This value type extern is basically C#'s `Span<T>` type, but for hxcpp. It is used to represent a contiguous region of memory, managed or unmanaged. You can use the `cpp.marshal.ViewExtensions` class to create views from haxe arrays, vectors, bytes, or pointers.

```haxe
final a = [ 0, 1, 2, 3 ];
final v = a.asView().slice(1, 2);

trace(v[0], v[1]); // prints 1, 2
````

There are then all sorts of operator for getting slices, copying, reinterpreting, etc. Since these are all stack only value types holding a pointer and a length they're nice and light weight.

### cpp.marshal.Marshal

This class provides a bunch of static functions for working with view types. Historically it has been quite annoying to deal with c-strings in haxe / hxcpp, but views and the marshal class makes these sort of operations easy.

```haxe
final s = "Hello, World";
final v = s.toWideCharView(); // cpp.marshal.View<cpp.Char16>

trace(v.toString());
```

The marshal class also provides functions for reading and writing any types from views, making it easy to pack objects into bytes.

```haxe
@:cpp.ValueType
extern class Point {
    var x : Int;
    var y : Int;

    function new(x:Int, y:Int):Void;
}

final a = [ 0, 0 ];
final v = a.asView().reinterpret();
final p = new Point(2, 4);

v.write(p);

trace(a);  // prints [ 2, 4 ];
```

This all works well and I'm really happy with it. I've been experimenting with removing untyped usage using this stuff and I've got a version of `haxe.io.Bytes` which is now untyped free. E.g. `getDouble` was,

`return untyped __global__.__hxcpp_memory_get_double(b, pos)`

and is now

`return this.asView().slice(pos).readFloat64()`

I've also got a wip implementation of the `sys.io.File` api and associated readers and writers which use the new managed type externs and views to completely remove the weak `Dynamic` typing and untyped usage there.

Aidan63 committed:
bit of code shuffling and checking for stack only types in classes and enums

Aidan63 committed:
store errors in one place

Aidan63 committed:
make view stack only and use size_t for length

Aidan63 committed:
Guard again promoting stack only value type in expressions

Aidan63 committed:
endian specific read and write functions, plus ref to view

Aidan63 committed:
view based FPHelper

Aidan63 committed:
Dodge pointer vs user var fusion issue

Aidan63 committed:
Change the public api to use int64 over size_t

Aidan63 committed:
throw when converting big view to haxe type

Aidan63 committed:
view based byte implementation

Aidan63 committed:
Merge branch 'development' into cpp-value-types

Aidan63 committed:
test using hxcpp fork

Aidan63 committed:
set docgen to use hxcpp fork as well

Aidan63 commented:
Few more updates.

I've changed the length of `cpp.marshal.View` to be `Int64`, this now means there's a nice way of working with large buffers. The `ToArray`, `ToVector`, and other conversion functions to haxe types which only use an `Int` for their length will throw if the view is too large.

The view type can safely hold managed and unmanaged pointers assuming it always lives on the stack, since hxcpp does conservative stack scanning and anything scanned on the stack is pinned. This requirement is now enforced by the compiler. I.e. the `cpp.marshal.View` now has the `StackOnly` flag which enforces this restriction.

E.g. the following sample

```haxe
import cpp.marshal.View;

using cpp.marshal.ViewExtensions;

function main() {
	final src  = [ 0, 0, 0, 0 ];
	final view = src.asView();

	final func = () -> {
		trace(view.length);
	}

	func();
}
```

Gives the following error, since view is captured in a closure which requires it being promoted to the heap.

```
[ERROR] Main.hx:7: characters 8-12

 7 |  final view = src.asView();
   |        ^^^^
   | CPP0011: Marshalling value type with the StackOnly flag cannot be promoted to the heap
```

This is not the only situation which requires heap promotion, class fields, enum constructors, nullable, etc, will all give the same error.

The `cpp.marshal.Marshal` class now has static functions for endian specific read and write functions along side the machine endianness ones.

Finally, I've added view based implementations of `haxe.io.FPHelper` and `haxe.io.Bytes` which removes all the untyped in the previous cpp implementations.

I've got more ideas for all this but I think this is good enough for now. I've changed the CI to use my hxcpp branch to show the tests all pass, not sure what order we need to merge this and the hxcpp side in.

Simn commented:
Glad to see you've reached a milestone here! These changes to Bytes.hx look great. 

I guess since both PRs depend on each other, we should just merge them at the same time and then hope that everything passes. So let's set the reference here back to HaxeFoundation/hxcpp and I'll go click the buttons.

tobil4sk commented:
One thing that I've found painful in the past is when externs assume the way a type will be used (i.e. only as a pointer, or only as a value), and if I want to change it it requires going through and modifying a bunch of extern declarations.

This has lead me to think that there might be benefit in encouraging extern definitions just be a pure declaration of the type fields. Then when using it, you could have the freedom to decide whether it should be a value or pointer. What are your thoughts?

I also wonder if it might be helpful to split this into smaller PRs with more focused diffs, as the large diff and long history will make it difficult to perform git bisects if any issues occur.

Aidan63 committed:
Revert "set docgen to use hxcpp fork as well"

This reverts commit fcd3c1add5b52fe7ea562a5b98a063557447ce5c.

Aidan63 committed:
Revert "test using hxcpp fork"

This reverts commit 31ecd0bf1d0c02da5f908131a07fe6bc66f94886.

Aidan63 commented:
> you could have the freedom to decide whether it should be a value or pointer

Potentially, the main thing I was aiming for with the `cpp.PointerType` is the common c / c++ pattern to pass around some context object which is always a pointer, and trying to do pointer specific things such as dereferencing is usually asking for trouble (lots of libs try to hide the pointer-ness behind a typedef).

There is support for the old pointer types in here as well, so for some edge case where you do need a pointer to a value type extern you can do that.

For the large diff, I'm not sure there's much I could do in this case, unfortunately these additions needed changes all over the place.

Simn commented:
I agree that large diffs can be a problem for the reasons you stated, but this is honestly not the worst PR I've seen in that regard, so I'm sure it will be fine.

Simn commented:
One question: isn't this supposed to increase the API level? The hxcpp PR has `#if (HXCPP_API_LEVEL >= 500)` changes, but 500 is already the version we're currently on.

Aidan63 commented:
My assumption has been that the hxcpp api level is related to the haxe version. e.g. HXCPP_API_LEVEL 500 is haxe 5.0.0 and since haxe 5 would be the first these additions are in it should be with that api level.
It does lead to these awkward situations in between haxe releases though, but I'm not sure hxcpp has version checking stuff to deal with that.

None of these changes modify the public signatures of any hxcpp stuff though, just new additions. The level checks are just there to prevent the new types being pulled into old haxe 4 builds in case it causes any issues.
